### PR TITLE
CAMEL-10988 - Improve performance of CaseInsensitiveMap

### PR DESCRIPTION
**Pull request by MarcinLachowicz**
https://issues.apache.org/jira/browse/CAMEL-10988

### CODE REVIEW

File: camel-core/src/main/java/org/apache/camel/util/GenericCaseInsensitiveMap.java [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+    }
+
+    public GenericCaseInsensitiveMap(Map<K, V> m) {
+        map = copy(m, new HashMap(m.size()));
+    }
+
+    protected Map<K, V> copy(Map<K, V> source, Map dest) {
+        for (Entry<K, V> entry : source.entrySet()) {
+            K key = entry.getKey();
+            K altKey;
```
Comment (aldettinger) [inline camel-core/src/main/java/org/apache/camel/util/GenericCaseInsensitiveMap.java]: I think we could get rid of the `altKey` variable and the corresponding else block as well.

**Comment by davsclaus** [comment]
We favor using the JDK libraries. And we dont fancy copying code in from other projects.

**Comment by MarcinLachowicz** [comment]
Even if it gives huge performance boost?

File: camel-core/src/main/java/org/apache/camel/util/GenericCaseInsensitiveMap.java (hunk 1 above)
Comment (MarcinLachowicz) [inline camel-core/src/main/java/org/apache/camel/util/GenericCaseInsensitiveMap.java]: Good catch, I've fixed it

**Comment by MarcinLachowicz** [comment]
@davsclaus 
I've prepared Microbenchmark for both actual CaseInsensitiveMap and new one. Benchmark executes the following scenario:
1. Initialize map with 10 predefined elements
2. Add 10 elements one by one
3. Remove elements that were added in step 1. one by one
4. Add 10 elements at one
Predefined elements keys are randomly generated String, which consists of 10 chars.
Microbenchmark code is attached. Just invoke mvn clean install in main dir then run java -jar target/case-insensitive-maps-benchmarks.jar

Here are the results from my machine:
```
Result "com.example.CaseInsensitiveMapsBenchmark.testCamelCaseInsensitiveMap":
  483599.059 ±(99.9%) 5915.112 ops/s [Average]
  (min, avg, max) = (474628.669, 483599.059, 492344.044), stdev = 5533.000
  CI (99.9%): [477683.947, 489514.171] (assumes normal distribution)

Result "com.example.CaseInsensitiveMapsBenchmark.testGenericCaseInsensitiveMap":
  622327.547 ±(99.9%) 17548.997 ops/s [Average]
  (min, avg, max) = (604195.743, 622327.547, 655964.494), stdev = 16415.343
  CI (99.9%): [604778.550, 639876.545] (assumes normal distribution)
```

As you can see, new CaseInsensitiveMap provides much better performance, which is crucial for high transnational systems that use Camel. Is it possible to proceed with this PR?

**Comment by davsclaus** [comment]
I create a ticket which explains this some more and an avenue for a solution
https://issues.apache.org/jira/browse/CAMEL-11149

This PR as it stands cannot be accepted. But the CAMEL-11149 ticket explains what can be done.

**Comment by davsclaus** [comment]
And btw sorry for the slow response, but the camel team is often busy and this ticket was not straightfoward due to that code copy.

You are welcome to look into CAMEL-11149 and contribute a custom component with this java-util implementation that uses their released JAR, eg no copy of code etc.

Then we can work on the SPI hook interface and how Camel discovers this automatic on startup so its easy to plugin and use

**Comment by MarcinLachowicz** [comment]
It's okay, I appreciate your feedback :) 
I gonna re-implement it as you suggested.

**Comment by davsclaus** [comment]
@MarcinLachowicz I wonder if you would like to contribute a PR with your JMH tests to Apache Camel. You can add the tests to: https://github.com/apache/camel/tree/master/tests/camel-jmh

**Comment by MarcinLachowicz** [comment]
@davsclaus sure, let me do this in next couple of days :)

**Comment by MarcinLachowicz** [comment]
@davsclaus I've just added my benchmarks here: https://github.com/apache/camel/pull/1767

Regards,
Marcin Lachowicz

**Pull request closed without merging**